compiler/next/lib/resolution/resolution-queries.cpp

```diff
@@ -902,112 +875,350 @@ typedSignatureInitial(Context* context,
+    }
+
+    ret = BasicClassType::get(context, c->id(), c->name(),
+                              parentClassType,
+                              insnFromBct, std::move(filteredSubs));
+
+  } else if (auto r = ad->toRecord()) {
+    const RecordType* insnFromRec = nullptr;
+    if (instantiatedFrom != nullptr) {
+      if (auto rec = instantiatedFrom->toRecordType())
+        insnFromRec = rec;
+      else
+        assert(false && "unexpected instantiatedFrom type");
+    }
+ 
+    ret = RecordType::get(context, r->id(), r->name(),
+                          insnFromRec, std::move(filteredSubs));
+
+  } else if (auto u = ad->toUnion()) {
+    const UnionType* insnFromUni = nullptr;
+    if (instantiatedFrom != nullptr) {
+      if (auto uni = instantiatedFrom->toUnionType())
+        insnFromUni = uni;
+      else
+        assert(false && "unexpected instantiatedFrom type");
+    }
+ 
+    ret = UnionType::get(context, u->id(), u->name(),
+                         insnFromUni, std::move(filteredSubs));
+
+  } else {
+    assert(false && "case not handled");
+  }
+
+  return ret;
+}
+
+// initedInParent is true if the decl variable is inited due to a parent
+// uast node.  This comes up for TupleDecls.
+static void helpSetFieldTypes(const ASTNode* ast,
+                              ResolutionResultByPostorderID& r,
+                              bool initedInParent,
+                              ResolvedFields& fields) {
+
+  if (auto var = ast->toVarLikeDecl()) {
+    bool hasDefaultValue = initedInParent || var->initExpression() != nullptr;
+    const ResolvedExpression& e = r.byAst(var);
+    fields.addField(var->name(), hasDefaultValue, var->id(), e.type());
+  } else if (auto mult = ast->toMultiDecl()) {
+    for (auto decl : mult->decls()) {
+      helpSetFieldTypes(decl, r, initedInParent, fields);
+    }
+  } else if (auto tup = ast->toTupleDecl()) {
+    bool hasInit = initedInParent || tup->initExpression() != nullptr;
+    for (auto decl : tup->decls()) {
+      helpSetFieldTypes(decl, r, hasInit, fields);
+    }
+  }
+
+  // no action needed for other types of Decls since they aren't fields.
+}
+
+// This visits the fields to resolve them and then
+// sets the types in the Type* passed.
+static void
+resolveAndSetFieldTypes(Context* context, const AggregateDecl* ad,
+                        ResolutionResultByPostorderID& r,
+                        Resolver& visitor,
+                        const CompositeType* t,
+                        ResolvedFields& fields) {
+
+  // visit the field declarations to resolve them
+  // and then set them in the type
+  for (auto child: ad->children()) {
+    child->traverse(visitor);
+    helpSetFieldTypes(child, r, /* initedInParent */ false, fields);
+  }
+
+  // finalize the field types to compute summary information
+  // like whether any was generic
+  fields.finalizeFields(context);
+}
+
+static const Type* const&
+initialTypeForTypeDeclQuery(Context* context, ID declId) {
```

Comment:
Why do we specifically call out the fact that this is a query here (in the name) and for `fieldsForTypeDeclQuery`?

Reply:
I answer this in answer to the next comment